Add Examples and documentation for pod_template_file feature (#10916)
mistercrunch/Airflow: docs/executor/kubernetes.rst

To troubleshoot issue with KubernetesExecutor, you can use ``airflow kubernetes generate-dag-yaml`` command.
This command generates the pods as they will be launched in Kubernetes and dumps them into yaml files for you to inspect.

.. _concepts:pod_template_file:

pod_template_file
#################

> As of Airflow 1.10.12, users can now use a pod_template_file setting in their airflow.cfg to form the basis of

Reviewer: ```suggestion
As of Airflow 1.10.12, users can now use a ``pod_template_file`` option in ``XXX`` section in their airflow.cfg to form the basis of
```

their KubernetesExecutor pods. This process is faster and easier to modify.

We include multiple examples of working pod operators below, but we would also like to explain a few necessary components

> if you want to customize your pod_template_files. As long as you have these components, every other element

Reviewer: ```suggestion
if you want to customize your Pod template files. As long as you have these components, every other element
```

in the template is
customizable.

1. Airflow will overwrite the base container image and the pod name

There are two points where airflow potentially overwrites the base image: in the airflow.cfg
or the ``pod_override`` (discussed below) setting. This value is overwitten to ensure that users do
not need to update multiple template files every time they upgrade their docker image. The other field
that Airflow overwrites is the ``pod.metadata.name`` field. This field has to be unique across all pods,
so we generate these names dynamically before launch.

It's important to note while Airflow overwrites these fields, they **can not be left blank**.
If these fields do not exist, kubernetes can not load the yaml into a Kuberentes V1Pod

> 2. Each airflow ``pod_template_file`` must have a container named ``base`` at the ``pod.spec.containers[0]`` position

Reviewer: ```suggestion
2. Each airflow ``pod_template_file`` must have a container named ``"base"`` at the ``pod.spec.containers[0]`` position
```

Airflow uses the ``pod_template_file`` by making certain assumptions about the structure of the template.
When airflow creates the worker pod's command, it assumes that the airflow worker container part exists

> at the beginning of the container array. It then assumes that the container is named "base"

Reviewer: ```suggestion
at the beginning of the container array. It then assumes that the container is named ``"base"``
```

> when it merges this pod with internal configs. Users are more than welcome to create

Reviewer: ```suggestion
when it merges this pod with internal configs. You are more than welcome to create
```
See:
https://developers.google.com/style/person
and 
> Address the reader directly. Use you, instead of the user or they. For more information, see [Second person and first person](https://developers.google.com/style/translation
).

https://developers.google.com/style/translation

To troubleshoot issue with KubernetesExecutor, you can use ``airflow kubernetes generate-dag-yaml`` command.
This command generates the pods as they will be launched in Kubernetes and dumps them into yaml files for you to inspect.

.. _concepts:pod_template_file:

pod_template_file

> #################

Reviewer 2: Needs a rst style heading

Author: @kaxil I'm not sure what you mean? This seems to be the same as the sections below it?

Reviewer 2: aah it works, yeah ignore me

> As of Airflow 1.10.12, you can now use the ``pod_template_file`` option in the ``kubernetes`` section
> of their airflow.cfg to form the basis of your KubernetesExecutor pods. This process is faster to execute
> and easier to modify.

Reviewer 2: ```suggestion
As of Airflow 1.10.12, you can now use the ``pod_template_file`` option in the ``kubernetes`` section
of the ``airflow.cfg`` file to form the basis of your KubernetesExecutor pods. This process is faster to execute
and easier to modify.
```

We include multiple examples of working pod operators below, but we would also like to explain a few necessary components
if you want to customize your template files. As long as you have these components, every other element
in the template is
customizable.

1. Airflow will overwrite the base container image and the pod name

> There are two points where airflow potentially overwrites the base image: in the airflow.cfg

Reviewer 2: ```suggestion
There are two points where Airflow potentially overwrites the base image: in the ``airflow.cfg``
```

or the ``pod_override`` (discussed below) setting. This value is overwitten to ensure that users do
not need to update multiple template files every time they upgrade their docker image. The other field
that Airflow overwrites is the ``pod.metadata.name`` field. This field has to be unique across all pods,
so we generate these names dynamically before launch.

It's important to note while Airflow overwrites these fields, they **can not be left blank**.

> If these fields do not exist, kubernetes can not load the yaml into a Kuberentes V1Pod

Reviewer 2: ```suggestion
If these fields do not exist, Kubernetes can not load the YAML into a Kubernetes V1Pod.
```

> 2. Each airflow ``pod_template_file`` must have a container named "base" at the ``pod.spec.containers[0]`` position

Reviewer 2: ```suggestion
2. Each Airflow ``pod_template_file`` must have a container named "base" at the ``pod.spec.containers[0]`` position
```

> Airflow uses the ``pod_template_file`` by making certain assumptions about the structure of the template.
> When airflow creates the worker pod's command, it assumes that the airflow worker container part exists
> at the beginning of the container array. It then assumes that the container is named ``base``
> when it merges this pod with internal configs. you are more than welcome to create
> sidecar containers after this required container.

Reviewer 2: ```suggestion
Airflow uses the ``pod_template_file`` by making certain assumptions about the structure of the template.
When airflow creates the worker pod's command, it assumes that the airflow worker container part exists
at the beginning of the container array. It then assumes that the container is named ``base``
when it merges this pod with internal configs. You are more than welcome to create
sidecar containers after this required container.
```

> With these requirements in mind, here are some examples of basic pod_template_file yaml files.

Reviewer 2: ```suggestion
With these requirements in mind, here are some examples of basic ``pod_template_file`` YAML files.
```

> pod_template_file using the dag_in_image setting:

Reviewer 2: ```suggestion
pod_template_file using the ``dag_in_image`` setting:
```

.. exampleinclude:: /../airflow/kubernetes_executor_templates/dags_in_image_template.yaml
    :language: yaml
    :start-after: [START template_with_dags_in_image]
    :end-before: [END template_with_dags_in_image]

> pod_template_file which stores DAGs in a persistentVolume:

Reviewer 2: ```suggestion
``pod_template_file`` which stores DAGs in a ``persistentVolume``:
```

.. exampleinclude:: /../airflow/kubernetes_executor_templates/dags_in_volume_template.yaml
    :language: yaml
    :start-after: [START template_with_dags_in_volume]
    :end-before: [END template_with_dags_in_volume]

> pod_template_file which pulls DAGs from git:

Reviewer 2: ```suggestion
``pod_template_file`` which pulls DAGs from git:
```
